Remove LazyStreamContext (2 out of 2)

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59377 Move CUDA-related stuff of TP agent to separate file
* #59376 Make CUDA serde support for TP agent pluggable
* #59375 Merge TensorPipe's CPU and CUDA channel registry
* **#59299 Remove LazyStreamContext (2 out of 2)**
* #59298 Remove LazyStreamContext (1 out of 2)
* #59297 Fix CUDA sync when switching streams in RPC tests
* #58753 Fix race condition in TP agent
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef

After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

lw committed:
Remove LazyStreamContext (2 out of 2)

After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -110,6 +113,48 @@ std::vector<c10::Stream> getCurrentStreamsForDevices(
+    const std::vector<torch::Tensor>& tensors) {
+  c10::optional<c10::impl::VirtualGuardImpl> impl;
+  size_t deviceCount = 0;
+  std::vector<bool> indexBitSet;
```
curious about this var name. It seems to be neither `bit` nor `set`? Or is it possible to use a real `std::bitset` here?

lw replied:
This is probably not the best capitalization, but the name was meant to be "a bitset of indices". It is conceptually identical to `std::bitset`, but if you notice `std::bitset` requires a _compile-time_ size. The doc for `std::bitset` says:
> If the size of the bitset is not known at compile time, std::vector<bool> may be used.

Hope this clarifies. I'll fix the capitalization.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -110,6 +113,48 @@ std::vector<c10::Stream> getCurrentStreamsForDevices(
+  }
+  std::vector<c10::Device> devices;
+  devices.reserve(deviceCount);
+  for (c10::DeviceIndex idx = 0; idx < indexBitSet.size(); idx++) {
```
lint failures look real.

lw replied:
The linter is correct "in principle" but here we know that the size of the vector will fit in a `DeviceIndex` because we used a `DeviceIndex` to resize the vector.

I'm starting to think that perhaps we should create a dedicated DeviceVector class, that can provide properly-typed accessors, can perhaps use small-vector optimizations, ...

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM. Stamp to unblock.

lw committed:
Update on "Remove LazyStreamContext (2 out of 2)"


After recent changes, LazyStreamContext had in fact always become eager, and was in fact equivalent to a vector of streams. So it makes more sense now to remove this abstraction and use a more self-descriptive type.

This PR migrates the TensorPipe agent. The previous PR migrated the RequestCallback internals.

Differential Revision: [D28789174](https://our.internmc.facebook.com/intern/diff/D28789174/)

[ghstack-poisoned]
